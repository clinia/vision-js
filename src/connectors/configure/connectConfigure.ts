import searchHelper, {
  SearchParameters,
  PlainSearchParameters,
  CliniaSearchHelper,
} from '@clinia/search-helper';
import {
  Renderer,
  RendererOptions,
  Unmounter,
  WidgetFactory,
} from '../../types';
import {
  createDocumentationMessageGenerator,
  isPlainObject,
  mergeSearchParameters,
  noop,
} from '../../lib/utils';

type Refine = (searchParameters: PlainSearchParameters) => void;

export interface ConfigureConnectorParams {
  /**
   * A list of search parameters
   * to enable when the widget mounts.
   */
  searchParameters: PlainSearchParameters;
}

export interface ConfigureRendererOptions<TConfigureWidgetParams>
  extends RendererOptions<TConfigureWidgetParams> {
  refine: Refine;
}

export type ConfigureRenderer<TConfigureWidgetParams> = Renderer<
  ConfigureRendererOptions<ConfigureConnectorParams & TConfigureWidgetParams>
>;

export type ConfigureWidgetFactory<TConfigureWidgetParams> = WidgetFactory<
  ConfigureConnectorParams & TConfigureWidgetParams
>;

export type ConfigureConnector = <TConfigureWidgetParams>(
  render?: ConfigureRenderer<TConfigureWidgetParams>,
  unmount?: Unmounter
) => ConfigureWidgetFactory<TConfigureWidgetParams>;

const withUsage = createDocumentationMessageGenerator({
  name: 'configure',
  connector: true,
});

function getInitialSearchParameters(
  state: SearchParameters,
  widgetParams: ConfigureConnectorParams
): SearchParameters {
  // We leverage the helper internals to remove the `widgetParams` from
  // the state. The function `setQueryParameters` omits the values that
  // are `undefined` on the next state.
  return state.setQueryParameters(
    Object.keys(widgetParams.searchParameters).reduce(
      (acc, key) => ({
        ...acc,
        [key]: undefined,
      }),
      {}
    )
  );
}

const connectConfigure: ConfigureConnector = (
  renderFn = noop,
  unmountFn = noop
) => {
  return widgetParams => {
    if (!widgetParams || !isPlainObject(widgetParams.searchParameters)) {
      throw new Error(
        withUsage('The `searchParameters` option expects an object.')
      );
    }

    type ConnectorState = {
      refine?: Refine;
    };

    const connectorState: ConnectorState = {};

    function refine(helper: CliniaSearchHelper): Refine {
      return (searchParameters: PlainSearchParameters) => {
        // Merge new `searchParameters` with the ones set from other widgets
        const actualState = getInitialSearchParameters(
          helper.state,
          widgetParams
        );
        const nextSearchParameters = mergeSearchParameters(
          actualState,
          new searchHelper.SearchParameters(searchParameters)
        );

        // Trigger a search with the resolved search parameters
        helper.setState(nextSearchParameters).search();

        // Update original `widgetParams.searchParameters` to the new refined one
        widgetParams.searchParameters = searchParameters;
      };
    }

    return {
      $$type: 'cvi.configure',

      init({ visionInstance, helper }) {
        connectorState.refine = refine(helper);

        renderFn(
          {
            refine: connectorState.refine,
            visionInstance,
            widgetParams,
          },
          true
        );
      },

      render({ visionInstance }) {
        renderFn(
          {
            refine: connectorState.refine!,
            visionInstance,
            widgetParams,
          },
          false
        );
      },

      dispose({ state }) {
        unmountFn();

        return getInitialSearchParameters(state, widgetParams);
      },

      getWidgetSearchParameters(state, { uiState }) {
        return mergeSearchParameters(
          state,
          new searchHelper.SearchParameters({
            ...uiState.configure,
            ...widgetParams.searchParameters,
          })
        );
      },

      getWidgetState(uiState) {
        return {
          ...uiState,
          configure: {
            ...uiState.configure,
            ...widgetParams.searchParameters,
          },
        };
      },
    };
  };
};

export default connectConfigure;
