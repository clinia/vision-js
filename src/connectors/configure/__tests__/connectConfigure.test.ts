import searchHelper, {
  SearchParameters,
  CliniaSearchHelper,
} from '@clinia/search-helper';

import { createSearchClient } from '../../../../test/mock/createSearchClient';
import connectConfigure from '../connectConfigure';
import {
  createInitOptions,
  createDisposeOptions,
} from '../../../../test/mock/createWidget';

describe('connectConfigure', () => {
  let helper: CliniaSearchHelper;

  beforeEach(() => {
    helper = searchHelper(createSearchClient(), '', {});
  });

  describe('Usage', () => {
    it('throws without searchParameters', () => {
      // @ts-ignore wrong options
      expect(() => connectConfigure()()).toThrowErrorMatchingSnapshot();
    });

    it('throws when you pass it a non-plain object', () => {
      expect(() => {
        // @ts-ignore wrong options
        connectConfigure()(new Date());
      }).toThrowErrorMatchingSnapshot();

      expect(() => {
        // @ts-ignore wrong options
        connectConfigure()(() => {});
      }).toThrowErrorMatchingSnapshot();
    });

    it('does not throw with a render function but without an unmount function', () => {
      expect(() => connectConfigure(jest.fn(), undefined)).not.toThrow();
    });

    it('with a unmount function but no render function does not throw', () => {
      // @ts-ignore wrong options
      expect(() => connectConfigure(undefined, jest.fn())).not.toThrow();
    });

    it('does not throw without render and unmount functions', () => {
      // @ts-ignore wrong options
      expect(() => connectConfigure(undefined, undefined)).not.toThrow();
    });
  });

  it('is a widget', () => {
    const render = jest.fn();
    const unmount = jest.fn();

    const customConfigure = connectConfigure(render, unmount);
    const widget = customConfigure({ searchParameters: {} });

    expect(widget).toEqual(
      expect.objectContaining({
        $$type: 'cvi.configure',
        init: expect.any(Function),
        render: expect.any(Function),
        dispose: expect.any(Function),
      })
    );
  });

  it('should apply searchParameters', () => {
    const makeWidget = connectConfigure();
    const widget = makeWidget({
      searchParameters: {
        perPage: 20,
      },
    });

    expect(
      widget.getWidgetSearchParameters!(new SearchParameters({}), {
        uiState: {},
      })
    ).toEqual(
      new SearchParameters({
        perPage: 20,
      })
    );
  });

  it('should apply searchParameters with a higher priority', () => {
    const makeWidget = connectConfigure();
    const widget = makeWidget({
      searchParameters: {
        perPage: 20,
      },
    });

    expect(
      widget.getWidgetSearchParameters!(
        new SearchParameters({
          perPage: 20,
        }),
        { uiState: { configure: { perPage: 20 } } }
      )
    ).toEqual(
      new SearchParameters({
        perPage: 20,
      })
    );

    expect(
      widget.getWidgetSearchParameters!(
        new SearchParameters({
          perPage: 20,
          queryType: 'prefix_last',
        }),
        { uiState: { configure: { perPage: 20 } } }
      )
    ).toEqual(
      new SearchParameters({
        perPage: 20,
        queryType: 'prefix_last',
      })
    );
  });

  it('should apply new searchParameters on refine()', () => {
    const renderFn = jest.fn();
    const makeWidget = connectConfigure(renderFn, jest.fn());
    const widget = makeWidget({
      searchParameters: {
        perPage: 20,
      },
    });

    helper.setState(
      widget.getWidgetSearchParameters!(
        new SearchParameters({
          // This facet is added outside of the widget params
          // so it shouldn't be overridden when calling `refine`.
          facets: ['brand'],
        }),
        { uiState: { configure: { perPage: 20 } } }
      )
    );
    widget.init!(createInitOptions({ helper }));

    expect(
      widget.getWidgetSearchParameters!(new SearchParameters({}), {
        uiState: { configure: { perPage: 20 } },
      })
    ).toEqual(
      new SearchParameters({
        perPage: 20,
      })
    );
    expect(helper.state).toEqual(
      new SearchParameters({
        perPage: 20,
        facets: ['brand'],
      })
    );

    const { refine } = renderFn.mock.calls[0][0];

    refine({ perPage: 3, facets: ['rating'] });

    expect(
      widget.getWidgetSearchParameters!(new SearchParameters({}), {
        uiState: { configure: { perPage: 3, facets: ['rating'] } },
      })
    ).toEqual(
      new SearchParameters({
        perPage: 3,
        facets: ['rating'],
      })
    );
    expect(helper.state).toEqual(
      new SearchParameters({
        perPage: 3,
        facets: ['brand', 'rating'],
      })
    );
  });

  it('should dispose only the state set by configure', () => {
    const makeWidget = connectConfigure();
    const widget = makeWidget({
      searchParameters: {
        perPage: 20,
      },
    });

    helper.setState(
      widget.getWidgetSearchParameters!(
        new SearchParameters({
          queryType: 'prefix_last',
        }),
        { uiState: { configure: { perPage: 20 } } }
      )
    );
    widget.init!(createInitOptions({ helper }));

    expect(
      widget.getWidgetSearchParameters!(new SearchParameters({}), {
        uiState: { configure: { perPage: 20 } },
      })
    ).toEqual(
      new SearchParameters({
        perPage: 20,
      })
    );
    expect(helper.state).toEqual(
      new SearchParameters({
        perPage: 20,
        queryType: 'prefix_last',
      })
    );

    const nextState = widget.dispose!(
      createDisposeOptions({ state: helper.state })
    );

    expect(nextState).toEqual(
      new SearchParameters({
        queryType: 'prefix_last',
      })
    );
  });

  describe('getWidgetState', () => {
    it('adds default parameters', () => {
      const makeWidget = connectConfigure();
      const widget = makeWidget({
        searchParameters: {
          perPage: 20,
        },
      });

      expect(
        widget.getWidgetState!({}, { helper, searchParameters: helper.state })
      ).toEqual({
        configure: { perPage: 20 },
      });
    });

    it('adds refined parameters', () => {
      const renderFn = jest.fn();
      const makeWidget = connectConfigure(renderFn);
      const widget = makeWidget({
        searchParameters: {
          perPage: 20,
        },
      });

      widget.init!(createInitOptions({ helper }));
      const { refine } = renderFn.mock.calls[0][0];

      refine({ analytics: false });

      expect(
        widget.getWidgetState!({}, { helper, searchParameters: helper.state })
      ).toEqual({
        configure: { analytics: false },
      });
    });

    it('adds refined (new) parameters', () => {
      const renderFn = jest.fn();
      const makeWidget = connectConfigure(renderFn);
      const widget = makeWidget({
        searchParameters: {
          perPage: 20,
        },
      });

      widget.init!(createInitOptions({ helper }));
      const { refine } = renderFn.mock.calls[0][0];

      refine({ query: 'unsafe toys' });

      expect(
        widget.getWidgetState!({}, { helper, searchParameters: helper.state })
      ).toEqual({
        configure: { query: 'unsafe toys' },
      });
    });

    it('merges with existing configuration', () => {
      const makeWidget = connectConfigure();
      const widget = makeWidget({
        searchParameters: {
          perPage: 20,
        },
      });

      expect(
        widget.getWidgetState!(
          { configure: { queryType: 'prefix_last' } },
          { helper, searchParameters: helper.state }
        )
      ).toEqual({
        configure: { perPage: 20, queryType: 'prefix_last' },
      });
    });

    it('overwrites existing configuration', () => {
      const makeWidget = connectConfigure();
      const widget = makeWidget({
        searchParameters: {
          perPage: 20,
        },
      });

      expect(
        widget.getWidgetState!(
          { configure: { perPage: 20 } },
          { helper, searchParameters: helper.state }
        )
      ).toEqual({
        configure: { perPage: 20 },
      });
    });
  });

  describe('getWidgetSearchParameters', () => {
    it('returns parameters set by default', () => {
      const makeWidget = connectConfigure();
      const widget = makeWidget({
        searchParameters: {
          perPage: 20,
        },
      });

      const sp = widget.getWidgetSearchParameters!(new SearchParameters(), {
        uiState: {},
      });

      expect(sp).toEqual(new SearchParameters({ perPage: 20 }));
    });

    it('returns parameters set by uiState', () => {
      const makeWidget = connectConfigure();
      const widget = makeWidget({ searchParameters: {} });

      const sp = widget.getWidgetSearchParameters!(new SearchParameters(), {
        uiState: {
          configure: {
            perPage: 20,
          },
        },
      });

      expect(sp).toEqual(new SearchParameters({ perPage: 20 }));
    });

    it('overrides parameters set by uiState', () => {
      const makeWidget = connectConfigure();
      const widget = makeWidget({
        searchParameters: {
          perPage: 20,
        },
      });

      const sp = widget.getWidgetSearchParameters!(new SearchParameters(), {
        uiState: {
          configure: {
            perPage: 20,
          },
        },
      });

      expect(sp).toEqual(new SearchParameters({ perPage: 20 }));
    });

    it('merges parameters set by uiState', () => {
      const makeWidget = connectConfigure();
      const widget = makeWidget({
        searchParameters: {
          queryType: 'prefix_last',
        },
      });

      const sp = widget.getWidgetSearchParameters!(new SearchParameters(), {
        uiState: {
          configure: {
            perPage: 20,
          },
        },
      });

      expect(sp).toEqual(
        new SearchParameters({
          perPage: 20,
          queryType: 'prefix_last',
        })
      );
    });

    it('merges with the previous parameters', () => {
      const makeWidget = connectConfigure();
      const widget = makeWidget({
        searchParameters: {
          disjunctiveFacets: ['brand'],
          disjunctiveFacetsRefinements: {
            brand: ['Apple'],
          },
        },
      });

      const sp = widget.getWidgetSearchParameters!(
        new SearchParameters({
          disjunctiveFacets: ['categories'],
          disjunctiveFacetsRefinements: {
            categories: ['Phone'],
          },
        }),
        {
          uiState: {},
        }
      );

      expect(sp).toEqual(
        new SearchParameters({
          disjunctiveFacets: ['categories', 'brand'],
          disjunctiveFacetsRefinements: {
            brand: ['Apple'],
            categories: ['Phone'],
          },
        })
      );
    });

    it('stores refined state', () => {
      const renderFn = jest.fn();
      const makeWidget = connectConfigure(renderFn);
      const widget = makeWidget({
        searchParameters: {
          queryType: 'prefix_last',
        },
      });

      widget.init!(createInitOptions({ helper }));
      const { refine } = renderFn.mock.calls[0][0];

      refine({ queryType: 'prefix_none' });

      const sp = widget.getWidgetSearchParameters!(new SearchParameters(), {
        uiState: {},
      });

      expect(sp).toEqual(
        new SearchParameters({
          queryType: 'prefix_none',
        })
      );
    });
  });
});
