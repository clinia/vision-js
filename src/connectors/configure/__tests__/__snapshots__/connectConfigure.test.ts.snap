// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`connectConfigure Usage throws when you pass it a non-plain object 1`] = `
"The \`searchParameters\` option expects an object.

See documentation: https://www.clinia.com/doc/api-reference/widgets/configure/js/#connector"
`;

exports[`connectConfigure Usage throws when you pass it a non-plain object 2`] = `
"The \`searchParameters\` option expects an object.

See documentation: https://www.clinia.com/doc/api-reference/widgets/configure/js/#connector"
`;

exports[`connectConfigure Usage throws without searchParameters 1`] = `
"The \`searchParameters\` option expects an object.

See documentation: https://www.clinia.com/doc/api-reference/widgets/configure/js/#connector"
`;
